File: Assignment 1/CS20BTECH11063_A1.py
def extendedEuclidean(a, b):
    A = a
    B = b
    x = 1
    y = 0
    u = 0
    v = 1
    
    while B != 0:
        q = A//B
        r = A%B
        tmp1 = x-u*q
        tmp2 = y-v*q
        
        A = B
        B = r
        x = u
        y = v
        u = tmp1
        v = tmp2
        
    return (x,y,A)

if __name__ == "__main__":
    # read input from file
    f = open("input-gcd.csv", "r")
    
    # read each line till end of file is reached
    for line in f:
        a, b = line.split(",")
        a = int(a)
        b = int(b)
        
        # compute gcd
        x, y, gcd = extendedEuclidean(a, b)
        
        # print the result
        # print("x = " + str(x) + ", y = " + str(y) + ", c = " + str(gcd))
        print("x = {}, y = {}, c = {}".format(x, y, gcd))
    
    f.close()

File: Assignment 1/CS20BTECH11063_A1.cpp
#include <iostream>
#include <gmpxx.h>
#include <bits/stdc++.h>

using namespace std;

vector<mpz_class> extendedEuclidean(mpz_class &a, mpz_class &b)
{
    mpz_class x, y, A, B, u, v;
    // A = max(a, b);
    // B = min(a, b);
    A = a;
    B = b;

    x = 1;
    y = 0;
    u = 0;
    v = 1;

    // Apply the extended euclidean algorithm
    while (B != 0)
    {
        mpz_class q = A / B;
        mpz_class r = A % B;
        mpz_class tmp1 = x - q * u;
        mpz_class tmp2 = y - q * v;
        A = B;
        B = r;
        x = u;
        y = v;
        u = tmp1;
        v = tmp2;
    }

    vector<mpz_class> result = {x, y, A};
    return result;
}

int main()
{
    // Opening the input file
    ifstream fin;
    fin.open("input-gcd.csv");

    if (!fin.is_open())
    {
        cout << "Error in opening file\n";
        return 1;
    }

    // ofstream fout;
    // fout.open("output-gcd.csv");

    // Read each line till end of file
    string line;
    while (getline(fin, line))
    {
        // cout<<line<<endl;
        // Parse the line from the file into a, b
        istringstream input_line(line);
        string a_str, b_str;
        getline(input_line, a_str, ',');
        getline(input_line, b_str, ',');

        // Convert a_str, b_str to mpz_class
        mpz_class a(a_str);
        mpz_class b(b_str);

        // Apply the extended euclidean algorithm
        vector<mpz_class> result = extendedEuclidean(a, b);
        mpz_class x = result[0];
        mpz_class y = result[1];
        mpz_class A = result[2];

        // Print the result
        cout << "x = " << x << ", y = " << y << ", c = " << A << "\n";
        // fout << x << "," << y << "," << A << "\n";
        // cout << "A = " << A << ", B = " << B << endl;
        // cout << "a = " << a << ", b = " << b << endl;
    }

    // fout.close();
    fin.close();

    return 0;
}
